graph_weather/data/gencast_dataloader.py

```diff
@@ -0,0 +1,384 @@
+        self.max_year = max_year
+
+        self.num_lon = len(self.data["longitude"].values)
+        self.num_lat = len(self.data["latitude"].values)
+        self.num_vars = len(self.data.keys())
+        self.pressure_levels = np.array(self.data["level"].values).astype(
+            np.float32
+        )  # Need them for loss weighting
+
+        self.time_step = time_step  # e.g. 12h steps correspond to time_step = 2 in a 6h dataset
+
+        self.atmospheric_features = atmospheric_features
+        self.single_features = single_features
+        self.static_features = static_features
+
+        self.clock_features = ["local_time_of_the_day", "elapsed_year_progress"]
+        # Lat and long will be added by the model itself in the graph
+
+    def __len__(self):
+        return sum(self.data["time.year"].values < self.max_year) - 2 * self.time_step
+
+    def __getitem__(self, item):
+        start = self.data.isel(time=[item, item + self.time_step])
+        end = self.data.isel(time=item + 2 * self.time_step)
+
+        # Stack atmospheric features for input
+        atmospheric_input_data = np.stack(
+            [
+                (start[f"{var}"].values - np.array(const.ERA5_MEANS[f"{var}"])[None, :, None, None])
+                / (np.array(const.ERA5_STD[f"{var}"])[None, :, None, None] + 0.0001)
+                for var in self.atmospheric_features
+            ],
+            axis=-1,
+        ).astype(np.float32)
+
+        atmospheric_input_data = einops.rearrange(
+            atmospheric_input_data, "t lev lon lat var -> t lon lat (var lev)"
+        )
+        atmospheric_input_data = np.nan_to_num(atmospheric_input_data)
+        assert not np.isnan(atmospheric_input_data).any()
+
+        # Stack single features for input
+        single_input_data = np.stack(
+            [
+                (start[f"{var}"].values - np.array(const.ERA5_MEANS[f"{var}"]))
+                / (np.array(const.ERA5_STD[f"{var}"]) + 0.0001)
+                for var in self.single_features
+            ],
+            axis=-1,
+        ).astype(np.float32)
+
+        single_input_data = np.nan_to_num(single_input_data)
+        assert not np.isnan(single_input_data).any()
+
+        # Stack the static features for input
+        static_input_data = np.stack(
+            [
+                (start[f"{var}"].values - np.array(const.ERA5_MEANS[f"{var}"]))
+                / (np.array(const.ERA5_STD[f"{var}"]) + 0.0001)
+                for var in self.static_features
+            ],
+            axis=-1,
+        ).astype(np.float32)
+        static_input_data = np.stack([static_input_data] * 2, axis=0)
+        static_input_data = np.nan_to_num(static_input_data)
+        assert not np.isnan(static_input_data).any()
+        assert (static_input_data[0] == static_input_data[1]).all()
+
+        # Stack the time features for input
+        day_of_year = start.time.dt.dayofyear.values / 365.0
+        sin_day_of_year = (
```

Comment:
Again this is a nit but I can see these transformations are used a few times and look very similar here, could be pulled out into a helper function where you just specify whether you want to use sin or cos to transform the day of the year

Reply:
You're absolutely right! I've just reorganized the functions to reuse them when possible (sin, cos, normalization, etc.)